### PR TITLE
Access query (DepKind) metadata through fields

### PR DESCRIPTION
**Pull request by cjgillot**
This refactors the access to query definition metadata (attributes such as eval always, anon, has_params) and loading/forcing functions to generate a number of structs, instead of matching on the DepKind enum. This makes access to the fields cheaper to compile. Using a struct means that finding the metadata for a given query is just an offset away; previously the match may have been compiled to a jump table but likely not completely inlined as we expect here.

A previous attempt explored a similar strategy, but using trait objects in #78314 that proved less effective, likely due to higher overheads due to forcing dynamic calls and poorer cache utilization (all metadata is fairly densely packed with this PR).

### CODE REVIEW

**Comment by cjgillot** [comment]
The bug should be fixed now.

**Comment by jyn514** [comment]
@cjgillot does this need another perf run? Can you explain what went wrong before and what your fix was?

**Comment by cjgillot** [comment]
Yes please.
I tried to manually inline `can_reconstruct_query_key`, and made copy-paste mistake reversing the whole logic.

File: compiler/rustc_middle/src/dep_graph/dep_node.rs [hunk 1]
```diff
@@ -58,18 +58,128 @@ use crate::traits::query::{
...
+use rustc_serialize::{Decodable, Decoder, Encodable, Encoder};
 use rustc_span::symbol::Symbol;
+use rustc_span::DUMMY_SP;
 use std::hash::Hash;
 
 pub use rustc_query_system::dep_graph::{DepContext, DepNodeParams};
 
+pub type DepKind = &'static DepKindStruct;
+pub struct DepKindStruct {
+    pub(super) index: DepKindIndex,
```
Comment (bjorn3) [inline compiler/rustc_middle/src/dep_graph/dep_node.rs:78]: Could the fields be public so you can define your own DepKind outside rustc?

**Comment by jyn514** [comment]
Perf looks like mostly minor regressions, with a small improvement on `match-stress-enum`. `rustc_incremental` went down a lot, but at the cost of hurting rustc_middle.

File: compiler/rustc_middle/src/dep_graph/dep_node.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_middle/src/dep_graph/dep_node.rs:78]: This whole file should be moved outside `rustc_middle` eventually. However, there should be an exact correspondence between `DepKind`s, `DepKindIndex` and `DEP_KINDS` for encoding / decoding.

**Comment by cjgillot** [comment]
I have not found the origin of the regression yet.
Making `rustc_incremental` faster makes no sense.

**Comment by cjgillot** [comment]
Removed `OnDiskCache`-related code, and added a few inlining directives.

**Comment by tgnottingham** [comment]
Does this increase the size of the `DepKind` inside `DepNode`s from 1 byte to 4/8? If so, that's a bit unfortunate, because there are some memory usage improvements we can make if `DepKind` is small (e.g. #78646).

**Comment by jyn514** [comment]
Well that's interesting - instruction count is ~the same, but wall time went down quite a bit. The bootstrap times went _up_ for some reason, though?

**Comment by cjgillot** [comment]
This updated version does not modify the `DepKind` type any more, to avoid losing memory improvements.
Matches on `DepKind` are replaced by a jump table using function pointers.

**Comment by JohnCSimon** [comment]
@cjgillot  - ping from triage, can you please resolve the merge conflict and then assign it back to s-waiting-on-review  - thank you

**Comment by Mark-Simulacrum** [comment]
Is there a high level explanation of the intent here? I would like to see a summary of the approaches tried so far, what this PR does; ideally, an explanation of some of the tradeoffs involved - I see a number of commits here and while the PR itself is not too big it does touch on a core part of the compiler.

Can you say more about this change? Is there an MCP perhaps?

I may not be the best reviewer here regardless, but I think these questions are ones whoever ends up reviewing will ask. I'm also not sure if some of the discussion might be hidden by GitHub -- there's been quite a bit of back and forth with rust-timer, at least...

Maybe someone from @rust-lang/wg-incr-comp should review this?

**Comment by cjgillot** [comment]
Hi @Mark-Simulacrum 

There are two general objectives:
1. make the query system as monomorphic as possible, so as to reduce rustc compilation time,
2. allow to split up the query system into several pieces talking to each other via the dep graph.

Point 1 is mostly guesswork.
I should submit a MCP for point 2, I just don't have enough of a design yet.
Refined objectives are to:
- declare pieces of the query system in other crates as rustc_middle,
- allow to drop one of those query sub-systems early and free the associated memory (I am thinking about dropping the HIR).

Having the description of all the dep kinds in an unique enum is a major stop in this work.
This PR aims to solve this one point by transforming the giant matches against DepKind,
which are closed-world representation of the query system,
into a jump table system where DepKinds can be added in a more flexible fashion.

File: compiler/rustc_middle/src/dep_graph/dep_node.rs [hunk 2]
```diff
@@ -248,6 +230,19 @@ pub mod dep_kind {
...
+                can_reconstruct_query_key: || {
+                    if contains_anon_attr!($($attrs)*) {
+                        return false;
+                    }
+
+                    $({
+                        erase!($tuple_arg_ty);
+                        return <query_keys::$variant<'_> as DepNodeParams<TyCtxt<'_>>>
+                            ::can_reconstruct_query_key()
+                    })*
```
Comment (Mark-Simulacrum) [inline compiler/rustc_middle/src/dep_graph/dep_node.rs]: This seems suspicious. The tuple_arg_ty doesn't seem like it needs to be recursed here, $variant is the only thing that matters, right? But previously it looks like the code this replaced did expand to the tuple arg's ability to be reconstructed... which is odd, because I would expect the first one to return given this expansion, while my expectation would be that we need all of the arguments to be reconstructible, not just the first one.

(The next commit in this PR seems to replace this code with the suggested lack of iteration; maybe that's correct but it seems like different behavior?)

This is the old version, before this PR, for reference:

```rust
                        DepKind :: $variant => {
                            // tuple args
                            $({
                                return <$tuple_arg_ty as DepNodeParams<TyCtxt<'_>>>
                                    ::can_reconstruct_query_key();
                            })*

                            true
                        }
```

**Review by Mark-Simulacrum — COMMENTED** [review]
GitHub is marking the comment as outdated but it isn't (just left on an intermediate commit).

I think this overall seems to be fine - I prefer direct declaration rather than the matches we had previously. However, it does seem to be a regression in the compiler's build times, so we'll need to investigate further before landing. It might be due to the added inlining that this is asking for - I am not entirely sure that it is necessary; but you've already run perf a bunch of times on this PR so I could be wrong.

I would also like to see an explanation (comments in code, at least) on each of the fields being added here, as well as an overall explanation of the framework being pursued and some docs on why.

Please also try to squash commits down - the first 10 or so commits are good, it was easy to review migration of one "field" at a time - but the added cosmetic and inlining commits (roughly last 4 or 5) make that harder, as they're modifying the migrations in the first few commits. I would rather re-review cohesive commits from scratch than trying to construct those myself; feel free to eagerly rebase and squash into those first several commits.

File: compiler/rustc_middle/src/dep_graph/dep_node.rs [hunk 3]
```diff
@@ -175,8 +203,16 @@ pub mod dep_kind {
                 const is_anon: bool = contains_anon_attr!($($attrs)*);
                 const is_eval_always: bool = contains_eval_always_attr!($($attrs)*);
 
+                #[inline(always)]
+                fn can_reconstruct_query_key() -> bool {
+                    !is_anon &&
```
Comment (Mark-Simulacrum) [inline compiler/rustc_middle/src/dep_graph/dep_node.rs:270]: Why are we checking is_anon here? it seems like that's done in the caller of this already?

That said I see we're calling this directly in future commits, but seems good to still unify on where that check should happen.

File: compiler/rustc_middle/src/dep_graph/dep_node.rs [hunk 4]
```diff
@@ -175,8 +203,16 @@ pub mod dep_kind {
...
+                    <query_keys::$variant<'_> as DepNodeParams<TyCtxt<'_>>>
+                        ::can_reconstruct_query_key()
```
Comment (Mark-Simulacrum) [inline compiler/rustc_middle/src/dep_graph/dep_node.rs:272]: I might be wrong, but this looks like it has different behavior to before - I'm not seeing a loop over the tuple_arg_ty here. Is that intentional?

File: compiler/rustc_middle/src/dep_graph/mod.rs [hunk 5]
```diff
@@ -155,7 +154,26 @@ impl<'tcx> DepContext for TyCtxt<'tcx> {
...
+        // that point we would already have re-done all the work we are trying to
+        // avoid doing in the first place.
+        // The solution is simple: Just explicitly call the `codegen_unit` query for
+        // each CGU, right after partitioning. This way `try_mark_green` will always
+        // hit the cache instead of having to go through `force_from_dep_node`.
+        // This assertion makes sure, we actually keep applying the solution above.
+        debug_assert!(
+            dep_node.kind != DepKind::codegen_unit,
+            "calling force_from_dep_node() on DepKind::codegen_unit"
+        );
```
Comment (Mark-Simulacrum) [inline compiler/rustc_middle/src/dep_graph/mod.rs:174]: Hm, it seems like we should inline this into the force_from_dep_node for codegen unit, right? otherwise it's going to expand to useless code for all other queries and get dynamically checked...

**Review by Mark-Simulacrum — COMMENTED** [review]
This is looking much better, thank you. Left a few more comments, but mostly minor things I hope.

**Comment by Mark-Simulacrum** [comment]
@bors try @rust-timer queue

I am a bit worried that the latest perf results from this PR don't have much effect on runtime behavior of the compiler, but do seem to slow down the compile times during bootstrap by ~4% overall. I do actually think this PR is a net improvement on the readability and ease of understanding the query system, though - and further optimizations may be easier due to it - but I am concerned about the compilation time increase. Can we perhaps drop the (somewhat orthogonal) inline(always) annotations I saw added in some of the commits - they seem like an obvious potential culprit.

File: compiler/rustc_middle/src/dep_graph/dep_node.rs [hunk 6]
```diff
@@ -308,8 +308,23 @@ impl DepNodeExt for DepNode {
...
+        true
+    }
+
+    fn to_fingerprint(&self, _: TyCtxt<'tcx>) -> Fingerprint {
+        Fingerprint::ZERO
+    }
+
+    fn recover(_: TyCtxt<'tcx>, _: &DepNode) -> Option<Self> {
+        Some(())
+    }
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_middle/src/dep_graph/dep_node.rs:493]: This change corresponds to the fallback case at former line 135 in `can_reconstruct_query`. However, this is actually dead code, as no query has key `()`.

File: compiler/rustc_middle/src/dep_graph/dep_node.rs (hunk 4 above)
Comment (cjgillot) [reply compiler/rustc_middle/src/dep_graph/dep_node.rs:272]: The loop over `tuple_arg_ty` was actually a loop over a single element. See rustc_macros/src/query.rs:480.
Directly using `query_keys` seemed simpler.

Comment (Mark-Simulacrum) [reply compiler/rustc_middle/src/dep_graph/dep_node.rs:272]: Ok, seems fine then. Thanks!

File: compiler/rustc_middle/src/dep_graph/dep_node.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mark-Simulacrum) [reply compiler/rustc_middle/src/dep_graph/dep_node.rs:493]: Ok, seems fine.

File: compiler/rustc_middle/src/dep_graph/mod.rs (hunk 5 above)
Comment (cjgillot) [reply compiler/rustc_middle/src/dep_graph/mod.rs:174]: As this is a `debug_assert`, I am not be worried. In this case, the discoverability of the comment is more important than the check itself.

**Comment by bjorn3** [comment]
Mixed small (<1%) improvements and regressions.

**Comment by cjgillot** [comment]
Change on bootstrap times is around 2%, possibly due to variance.

**Comment by Mark-Simulacrum** [comment]
@cjgillot Am I correct that you believe you've resolved all of the comments I've left?

**Comment by cjgillot** [comment]
@Mark-Simulacrum Yes. I rebased and added 3 commits to address your comments.

**Comment by Mark-Simulacrum** [comment]
r=me with the last 3 commits squashed into the appropriate prior commits

File: compiler/rustc_middle/src/dep_graph/dep_node.rs [hunk 7]
```diff
@@ -103,79 +197,151 @@ macro_rules! contains_eval_always_attr {
...
+        try_load_from_on_disk_cache: |_, _| {},
+    };
+
+    pub const TraitSelect: DepKindStruct = DepKindStruct {
+        has_params: false,
+        is_anon: true,
+        is_eval_always: false,
+
+        can_reconstruct_query_key: || true,
+        force_from_dep_node: |_, _| false,
```
Comment (tgnottingham) [inline compiler/rustc_middle/src/dep_graph/dep_node.rs:234]: `TraitSelect` and `CompileCodegenUnit` used to `bug!` on `force_from_dep_node`, as with `Null` and `CrateMetadata`. Was this change intentional? Sorry if this is obvious--I haven't read the whole PR.

Comment (cjgillot) [reply compiler/rustc_middle/src/dep_graph/dep_node.rs:234]: Actually, `force_from_dep_node` bailed out earlier (line 167), since those DepKinds are either anonymous, or with unreconstructible key. I was aiming at keeping the exact same behaviour, but panicking should be fine.

Comment (tgnottingham) [reply compiler/rustc_middle/src/dep_graph/dep_node.rs:234]: Gotcha, I see now.
